### PR TITLE
Prep unit tests for React 18 upgrade

### PR DESCRIPTION
**Pull request by sloansparger**
Related to #43375

### Description

Pulls out all the changes required to make upgrading to React 18 + React Testing Library 15 possible but happen to be backwards compatible with React 17 + RTL 12. The changes were extracted from #41577 with the goal that we can backport this PR to minimize the diff between `master` and the current release branch, while also keeping our React 18 upgrade branch as small as possible.

Most changes fall into a few categories:
1. Use `await userEvent.click(el)` over `el.click()` (or similar event methods)
2. Use `act` where needed (generally for reacting to route changes or dispatched redux events)
3. Use `find` methods over `get` (I _think_ to account for async scheduling of the React 18 renderer)

### How to verify

- Unit tests should pass

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/admin/permissions/test/DatabasesPermissionsPage/DatabasesPermissionsPage.unit.spec.tsx [hunk 1]
```diff
@@ -64,10 +64,9 @@ const setup = async () => {
...
 
 const editDatabasePermission = async () => {
-  const permissionsSelectElem =
-    screen.getAllByTestId("permissions-select")[
-      NATIVE_QUERIES_PERMISSION_INDEX
-    ];
+  const permissionsSelectElem = (
+    await screen.findAllByTestId("permissions-select")
+  )[NATIVE_QUERIES_PERMISSION_INDEX];
   fireEvent.click(permissionsSelectElem);
```
Comment (npfitz) [inline frontend/src/metabase/admin/permissions/test/DatabasesPermissionsPage/DatabasesPermissionsPage.unit.spec.tsx:70]: Should we update this to userEvent?

File: frontend/src/metabase/containers/AdHocQuestionLoader.unit.spec.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ describe("AdHocQuestionLoader", () => {
     const q = Question.create({ databaseId: 1, tableId: 2 });
     const questionHash = ML_Urls.getUrl(q).match(/(#.*)/)[1];
 
-    render(
+    await render(
```
Comment (npfitz) [inline frontend/src/metabase/containers/AdHocQuestionLoader.unit.spec.js:25]: does this actually work?

File: frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js [hunk 3]
```diff
@@ -291,7 +294,7 @@ describe("DataSelector", () => {
     fireEvent.click(screen.getByText("Sample Database"));
```
Comment (npfitz) [inline frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js:294]: Can we replace these with userEvent.click as well? (lines 299, 329, 332 etc)

File: frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js [hunk 4]
```diff
@@ -418,7 +421,7 @@ describe("DataSelector", () => {
     ).toBeInTheDocument();
   });
 
-  it("should move between selected multi-schema dbs", () => {
+  it("should move between selected multi-schema dbs", async () => {
```
Comment (npfitz) [inline frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js:424]: It's not clear why we made this (And line 405) async

Comment (npfitz) [reply frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js:424]: Also, there are some occurrences of fireEvent in these tests

File: frontend/src/metabase/ui/components/inputs/MultiAutocomplete/MultiAutocomplete.unit.spec.tsx [hunk 5]
```diff
@@ -140,15 +140,15 @@ describe("MultiAutocomplete", () => {
 
   it("should accept quote-delimited values containing commas when pasting", async () => {
     const { input, onChange } = setup({ data: EXAMPLE_DATA });
-    input.focus();
+    act(() => input.focus());
```
Comment (npfitz) [inline frontend/src/metabase/ui/components/inputs/MultiAutocomplete/MultiAutocomplete.unit.spec.tsx:143]: calling `userEvent.click()` might be an alternative here that is more "what a user would do".

**Review by npfitz — APPROVED** [review]
This is awesome. Nothing here is a blocker, just thoughts as I read through.

File: frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js (hunk 4 above)
Comment (sloansparger) [reply frontend/src/metabase/query_builder/components/DataSelector/DataSelector.unit.spec.js:424]: honestly it's just annoying to have to change it to async when you want to await. i'd love to introduce a lint rule to force folks to always using async callbacks for test fns.

**Pull request merged**